OpenTerm/Controller/ViewController.swift

```diff
@@ -339,6 +357,261 @@ extension ViewController: UIDocumentPickerDelegate {
+                 *  Get all files that are in the bookmarks directory.
+                 */
+                let bookmarkFiles = try FileManager.default.contentsOfDirectory(atPath: bookmarkDirectoryURL.path)
+                
+                /**
+                 *  The array that will hold the obtained URLs. This will be returned eventually.
+                 */
+                var bookmarkURLs = [URL]()
+                
+                /**
+                 *  Iterate all bookmark filenames.
+                 */
+                for bookmarkFileName in bookmarkFiles {
+                    
+                    /**
+                     *  Get the url of the current bookmark file.
+                     */
+                    let bookmarkDataURL = bookmarkDirectoryURL.appendingPathComponent(bookmarkFileName)
+                    
+                    /**
+                     *  We try to load the bookmark from the file.
+                     */
+                    let loadedBookmark = try URL.bookmarkData(withContentsOf: bookmarkDataURL)
+                    
+                    /**
+                     *  This variable will indicate whether the bookmark is stale.
+                     *  If the bookmark is stale we create a new bookmark with the
+                     *  obtained URL and save the new one instead of the old one.
+                     */
+                    var isStale = true
+                    
+                    /**
+                     *  Try to obtain the URL from the bookmark.
+                     */
+                    if let loadedBookmarkURL = try URL(resolvingBookmarkData: loadedBookmark, bookmarkDataIsStale: &isStale) {
+                        
+                        /**
+                         *  If the bookmark is stale, we create a new bookmark
+                         *  from the obtained URL.
+                         */
+                        if isStale {
+                            do {
+                                try self.saveBookmarkURL(url: loadedBookmarkURL)
+                            }
+                        }
+                        
+                        /**
+                         *  Append the loaded URLs.
+                         */
+                        bookmarkURLs.append(loadedBookmarkURL)
+                    }
+                }
+                
+                return bookmarkURLs
+            } catch {
+                return []
+            }
+        }
+        
+        return []
+    }
+    
+    /// When the bookmark view controller did select a bookmark, we change the
+    /// current directory to the bookmarked url.
+    /// - Note: Only urls that were saved as bookmarks will work.
+    ///
+    /// - Parameter bookmarkURL: The bookmark that was selected.
+    func didSelectBookmarkURL(bookmarkURL: URL) {
+        
+        /**
+         *  Access the bookmarked URL
+         */
+        _ = bookmarkURL.startAccessingSecurityScopedResource()
```

Comment:
I believe that Apple recommends that this call needs to be balanced with a call to `stopAccessingSecurityScopedResource()`